logging: Fix runtime message creation in user-mode thread

cjwinklhofer opened the pull request:
### Description

The runtime message creation (CONFIG_LOG_ALWAYS_RUNTIME=y) in a user-mode thread raises a MPU violation, e.g. call LOG_INF("Test") in a user-mode thread. The function 'z_log_msg_runtime_vcreate' runs in user mode but works with Kernel data. Hence, create the cbprintf package on the stack (of the user mode thread) and pass it further to the syscall 'z_log_msg_static_create'.

Moreover, if CONFIG_NO_OPTIMIZATIONS=y is active, a user-mode thread is not allowed to access the logging module data (e.g. __log_level, __log_current_const_data). The macro LOG_MODULE_REGISTER now has a third argument to specify a partition for this logging module data so that the user-thread can access them.

### Discussion

In case CONFIG_NO_OPTIMIZATIONS=n, the access to the logging module data is optimized out and the standard 'LOG_MODULE_REGISTER(main);' is sufficient. I am not sure, if the logging module should implicitly define the partition (e.g. k_log_partition) only when CONFIG_NO_OPTIMIZATIONS=y and annotate the module data, instead of passing the partition via LOG_MODULE_REGISTER.

### Example

Extend the example with logging: _samples/userspace/hello_world_user/_

```
CONFIG_USERSPACE=y
CONFIG_ASSERT=y

CONFIG_LOG=y
CONFIG_NO_OPTIMIZATIONS=y
CONFIG_LOG_ALWAYS_RUNTIME=y
CONFIG_DEBUG=y
```

```
#include <zephyr/kernel.h>
#include <zephyr/logging/log.h>
#include <zephyr/logging/log_msg.h>
#include <zephyr/sys/libc-hooks.h>
#include <stdio.h>

#define USER_STACKSIZE	2048

#ifndef CONFIG_USERSPACE
#error This sample requires CONFIG_USERSPACE.
#endif

K_APPMEM_PARTITION_DEFINE(my_log_partition);

/* Sufficient for CONFIG_NO_OPTIMIZATIONS=n, otherwise partition is required */
/* LOG_MODULE_REGISTER(main); */
LOG_MODULE_REGISTER(main, CONFIG_LOG_DEFAULT_LEVEL, my_log_partition);

static struct k_mem_domain app_domain;
static struct k_mem_partition *app_parts[] = {
	&z_libc_partition,
	&my_log_partition
};

struct k_thread user_thread;
K_THREAD_STACK_DEFINE(user_stack, USER_STACKSIZE);

static void user_function(void *p1, void *p2, void *p3)
{
	LOG_INF("Hello World from %s (%s)\n",
	       k_is_user_context() ? "UserSpace!" : "privileged mode.",
	       CONFIG_BOARD);
	__ASSERT(k_is_user_context(), "User mode execution was expected");
}

int main(void)
{
	LOG_INF("Test Logging");

	k_tid_t tid = k_thread_create(&user_thread, user_stack, USER_STACKSIZE,
			user_function, NULL, NULL, NULL,
			-1, K_USER, K_FOREVER);

	k_mem_domain_init(&app_domain, ARRAY_SIZE(app_parts), app_parts);
	k_mem_domain_add_thread(&app_domain, tid);
	k_thread_start(tid);

	return 0;
}
```

nordic-krch commented:
I'm not that familiar with userspace and I don't use it but changes in the function look ok. I would prefer that `LOG_MODULE_REGISTER` is kept as is and some internal partition is used. Most modules will not have that 3rd parameter added so they won't work when used in userspace, also it is only for case when runtime message creation is used which is not the default one.

cjwinklhofer commented:
> I'm not that familiar with userspace and I don't use it but changes in the function look ok. I would prefer that `LOG_MODULE_REGISTER` is kept as is and some internal partition is used. Most modules will not have that 3rd parameter added so they won't work when used in userspace, also it is only for case when runtime message creation is used which is not the default one.

Ok thanks - I also find it better to left the arguments of LOG_MODULE_REGISTER untouched. Added the partition 'k_log_partition' when userspace and runtime message creation is active so that a user thread can have access.

nordic-krch commented:
looks ok. Can you try to restore userspace test if it is possible? It was removed here: https://github.com/zephyrproject-rtos/zephyr/commit/aa2b72d32af01e1b667eefcd92bf64ca135a1dc4
Maybe with your changes it will pass now.

cjwinklhofer commented:
> looks ok. Can you try to restore userspace test if it is possible? It was removed here: [aa2b72d](https://github.com/zephyrproject-rtos/zephyr/commit/aa2b72d32af01e1b667eefcd92bf64ca135a1dc4) Maybe with your changes it will pass now.

Yes, I will try to run this test.

cjwinklhofer commented on tests/subsys/logging/log_core_additional/testcase.yaml:
```diff
@@ -14,3 +14,11 @@ tests:
     extra_args: CONF_FILE=log_thread.conf
     integration_platforms:
       - native_sim
+  logging.log_user:
```
With CONFIG_LOG_ALWAYS_RUNTIME=y, the test 'tests/subsys/logging/log_core_additional/src/log_test_user.c' fails with an MPU error. Probably the k_log_partition must be assigned - however, I currently do not know how to add the test-thread (ZTEST_USER) to this partition. Will check this.

cjwinklhofer replied:
Ok, this can be done in the setup of the test-suite, added the partitions. My local tests looked good, lets see :)

cjwinklhofer commented:
There is a compile error on the RISC platform due to incompatible types for '__text_region_start':
```
/__w/zephyr/zephyr/arch/riscv/core/stacktrace.c: In function 'in_text_region':
/__w/zephyr/zephyr/arch/riscv/core/stacktrace.c:93:26: error: conflicting types for '__text_region_start'; have 'uintptr_t' {aka 'long unsigned int'}
   93 |         extern uintptr_t __text_region_start, __text_region_end;
      |                          ^~~~~~~~~~~~~~~~~~~
In file included from /__w/zephyr/zephyr/include/zephyr/app_memory/app_memdomain.h:9,
                 from /__w/zephyr/zephyr/include/zephyr/logging/log.h:15,
                 from /__w/zephyr/zephyr/arch/riscv/core/stacktrace.c:11:
/__w/zephyr/zephyr/include/zephyr/linker/linker-defs.h:142:13: note: previous declaration of '__text_region_start' with type 'char[]'
  142 | extern char __text_region_start[];
```

Would it not make sense to also declare __text_region_start in _arch/riscv/core/stacktrace.c_ as:
`extern char __text_region_start[], __text_region_end[];`
or remove them at all and only include  'include/zephyr/linker/linker-defs.h', which already contains them.

I did a grep and there are also other files that uses `extern uintptr_t __text_region_start`, so maybe uintptr_t is the correct way - I am not sure how to proceed.

cjwinklhofer commented:
Think I figured it out, both of them are correct and can be used for a symbol from the linker script (e.g. __text_region_start). Either by uintptr_t and the address operator &, or by declaring it as a char-array and using that without the address operator. However, they should be consistent, I will check their usage and in case it makes sense, prepare an additional commit that unifies them.

cjwinklhofer committed:
logging: Fix runtime message creation in user-mode thread

The runtime message creation (CONFIG_LOG_ALWAYS_RUNTIME=y) in a user-
mode thread raises a MPU violation, e.g. call LOG_INF("Test") in a
user-mode thread.

The function 'z_log_msg_runtime_vcreate' runs in user mode but works
with Kernel data. Hence, create the cbprintf package on the stack (of
the user mode thread) and pass it further to the syscall
'z_log_msg_static_create'.

Signed-off-by: Christoph Winklhofer <cj.winklhofer@gmail.com>

cjwinklhofer committed:
logging: Assign module data to memory partition 'k_log_partition'

The logging module data is assigned to the memory partition
'k_log_partition' so that a user mode thread can access this data (see
k_mem_domain_add_thread()). The 'k_log_partition' is created when:
- CONFIG_USERSPACE=y
- CONFIG_LOG_ALWAYS_RUNTIME=y

The option CONFIG_NO_OPTIMIZATIONS=y forces the logging module to use
the runtime message creation CONFIG_LOG_ALWAYS_RUNTIME=y. This raises a
MPU violation when logging is used in a user mode thread since this
thread is not allowed to access the module data (e.g. __log_level,
__log_current_const_data).

Note that the user mode thread may also require access to the partition
'z_libc_partition'.

Signed-off-by: Christoph Winklhofer <cj.winklhofer@gmail.com>

cjwinklhofer committed:
arch: Unify declaration of text region

The declaration for the text region '__text_region_start' and
'__text_region_end' should be consistent with the declaration in the
include file 'zephyr/linker/linker-defs.h'.

The local declarations are:
  extern uintptr_t __text_region_start, __text_region_end;
whereas 'linker_defs.h' declares them as:
  extern char __text_region_start[];
  extern char __text_region_end[];

This may result in conflicting types when 'linker_defs.h' is indirectly
included. Hence, remove the local declarations.

Signed-off-by: Christoph Winklhofer <cj.winklhofer@gmail.com>

cjwinklhofer committed:
tests: logging: restore userspace testing

Restore userspace testing that was removed with commit 'aa2b72d32af
(tests: logging: remove userspace testing, 2024-04-10)' and reported in
the issue #49213.

For the tests running with threads in user space (with ZTEST_USER),
the memory partitions are configured in the test-suite setup.

Signed-off-by: Christoph Winklhofer <cj.winklhofer@gmail.com>

ycsin reviewed (approved):
Changes to `arch/riscv` and `subsys/profiling/perf` look fine

dleach02 commented:
With this merge I'm getting failure:

west build -p -b ip_k66f/mk66f18 tests/subsys/logging/log_core_additional -T logging.log_user

cjwinklhofer commented:
> With this merge I'm getting failure:
> 
> west build -p -b ip_k66f/mk66f18 tests/subsys/logging/log_core_additional -T logging.log_user
> 
> The Segger ip_k66f board has a default config that turns on RTT. That in turn wants to enable PRINTK which this test case seems to have turned off?

Thanks! The reason was that without _CONFIG_LOG_PRINTK=n,_ the automatic tests did not build, since USERSPACE=y also implies LOG_PRINTK=n. However, I think with this PR, LOG_PRINTK can also be called from userspace and the dependency '!USERSPACE' in _subsys/logging/Kconfig.processing_ could be removed:

```
config LOG_PRINTK
bool "Process printk messages"
#depends on !USERSPACE
default y if PRINTK
```
